### PR TITLE
Literate Haskell Semantics - Syntax

### PR DESCRIPTION
**Pull request by Gabriel439**
Related to: https://github.com/dhall-lang/dhall-lang/issues/959

This is the first step towards an executable reference implementation by
creating a Haskell project which contains the Haskell type for the AST.

The Haskell project is built by CI, in order to catch any errors.

Note that the long-term goal is to standardize a normalization-by-evaluation
approach for the semantics in order to make it easier for
implementers to build efficient implementations.  However, in the short-term
I'm translating the semantics as literally as possible to Haskell
code and then I'll do a separate pass later on to switch to
normalization-by-evaluation.

### CODE REVIEW

**Comment by Nadrieril** [comment]
Yay thanks for doing this! Could you add a note in the README explaining how to compile this?

**Review by Nadrieril — APPROVED** [review]

File: standard/standard.cabal [hunk 1]
```diff
@@ -0,0 +1,19 @@
+cabal-version:       >=1.10
+name:                standard
+version:             1.0.0
+synopsis:            Literate Haskell semantics for Dhall
+description:         The Dhall semantics are written in literate Haskell,
+                     meaning that the logical jugdments are written side-by-side
+                     with the equivalent Haskell code in order to minimize
+                     ambiguity.
```
[the remark this answers is not in the document]
Comment (Gabriel439) [reply standard/standard.cabal:8]: I just put up a separate PR for the Agda version of this: https://github.com/dhall-lang/dhall-lang/pull/1103

**Comment by Gabriel439** [comment]
I'll let people comment on whether they prefer Haskell or Agda for this purpose.  If nobody comments then I'll throw in my own preference after a few days

File: standard/standard.cabal (hunk 1 above)
[the remark this answers is not in the document]
Comment (mheiber) [reply standard/standard.cabal:8]: Thank you for doing all this work in response to my small comment

**Comment by Gabriel439** [comment]
I'll insert my own preference now since a few people have already commented on the issues I believe are relevant to the comparison.  My suggestion is:

* go with just the Haskell implementation for now
* later on, possibly replace the natural deduction notation in the standard with Agda (but still keep the Haskell implementation)

My main reason is that the original motivation for doing this work is to make it as easy as possible for new language bindings to translate the semantics to their respective languages.  In other words, this implementation needs to act like a "Rosetta stone" for people without a PLT background who need to understand the correspondence between natural deduction and code.

I think Haskell is slightly better suited in that regard because I intend to use a highly simplified subset of functional programming to implement the semantics, and Haskell is idiomatically better suited to that subset than Agda.  If we go with Agda, there will be pressure to use Agda's more advanced features (like mixfix notation), which will interfere with the above goal of "explaining" the natural deduction notation.

There are also technical issues that limit the Agda implementation.  @Nadrieril pointed out that we will likely need to FFI out to a non-trivial amount of Haskell functionality, and maintaining a mixed Haskell + Agda implementation might be too much complexity for contributors to the standard.

Also, one last reason I prefer Haskell over Agda is that Haskell has broader editor support, whereas Agda is somewhat tightly coupled to `emacs`.  I'd like to make it as easy as possible for contributors to amend the standard.

**Comment by TristanCacqueray** [comment]
Thank you @Gabriel439 for proposing such reference implementation here, it seems like this would also be very useful to validate changes to the Dhall standard. Agda looks ideal, but if it's going to require Haskell, then starting with Haskell sounds better. I'm looking forward how Dhall can be implemented with highly simplified fp!

**Pull request merged**